Correlator: support for running barrel TM18

gpetruc opened the pull request:
This PR extends the emulators of the correlator layer 1 to support running the barrel at tmux 18 (albeit for now only on serenity boards)

There are no changes in the default correlator layer 1 configuration that is used in the L1 menu, which runs an idealized version of the barrel at tmux6.

Having the tmux18 version in CMSSW is however useful to do studies of this possible architecture and prepare pattern files for single and multi board tests at b904.

If this looks ok to you, I'll make the corresponding PR to CMSSW master this week

@cerminar

triggerDoctor commented:
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

Attempts to compile this PR succeeded!

|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -j 8`|

triggerDoctor commented:
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found no issues with the code checks!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-checks && scram b -k -j 8 code-checks`|

I found no issues with the headers!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 check-headers`|

triggerDoctor commented:
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found 1 files that did not meet formatting requirements:
 - L1Trigger/L1TTrackMatch/plugins/L1TrackJetClustering.h

Please run `scram b code-format` to auto-apply code formatting
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-format-all`|

gpetruc commented:
Hello triggerDoctor, 

I can fix it but it's not in my package so it must have entered in another PR.

Giovanni

> Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.
> 
> I found 1 files that did not meet formatting requirements:
> 
>     * L1Trigger/L1TTrackMatch/plugins/L1TrackJetClustering.h
> 
> 
> Please run `scram b code-format` to auto-apply code formatting
> Info 	Value
> return code 	`0`
> command 	`eval `scramv1 runtime -sh` && scram b -k -j 8 code-format-all`

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/interface/regionizer/multififo_regionizer_elements_ref.h:
```diff
@@ -76,6 +77,64 @@ namespace l1ct {
+      }
+      void readNewEvent() {
+        iread_ = 1 - iread_;
+        //std::cout << "Starting to read on " << iread_ << " (size " << items_[iread_].size() << ", other size " << items_[1-iread_].size() << ", writing " << iwrite_ << ")" << std::endl;
```
Could you pull the commented `cout`'s out please?

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/interface/regionizer/multififo_regionizer_elements_ref.h:
```diff
@@ -183,4 +242,57 @@ namespace l1ct {
+      items_[iwrite_].push_back(t);
+      items_[iwrite_].back().hwEta += etaShift_;
+      items_[iwrite_].back().hwPhi += phiShift_;
+    } else {
+      // uncommenting the message below may be useful for debugging
+      //dbgCout() << "WARNING: sector buffer is full for " << typeid(T).name() << ", pt = " << t.intPt()
+      //          << ", eta = " << t.intEta() << ", phi = " << t.intPhi() << "\n";
+    }
```
Is this debug still necessary? If so could you remove the commented stuff and put in `LogDebug` instead?

gpetruc replied:
dbgCout() gets compiled to the CMS MessageLogger when compiled within CMSSW, and to cout when compiled standalone (this emulator is also used from within Vitis HLS testbenches, where the MessageLogger is not avaiable).

I'd prefer to keep it as is; this code was already in CMSSW ( https://github.com/cms-sw/cmssw/blob/master/L1Trigger/Phase2L1ParticleFlow/interface/regionizer/buffered_folded_multififo_regionizer_ref.h#L128-L130 ), I just refactored to a different place.

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/interface/regionizer/multififo_regionizer_elements_ref.icc:
```diff
@@ -41,6 +41,11 @@ void l1ct::multififo_regionizer::maybe_push<l1ct::TkObjEmu>(const l1ct::TkObjEmu
+  bool isGood =
+      (nfifos == 1 || nfifos == 2 || nfifos == 3 || nfifos == 4 || nfifos == 6 || nfifos == 8 || nfifos == 12);
+  if (!isGood) {
+    dbgCerr() << "Error, created regionizer for nfifos == " << nfifos << ", not supported." << std::endl;
```
If this isn't one of the usual CMSSW Logging streams, could this be removed in favor of `LogError`?

gpetruc replied:
as above, `dbgCerr()` goes to MessageLogger. See https://github.com/cms-sw/cmssw/blob/master/L1Trigger/Phase2L1ParticleFlow/interface/dbgPrintf.h

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/src/regionizer/middle_buffer_multififo_regionizer_ref.cpp:
```diff
@@ -0,0 +1,632 @@
+
+void l1ct::MiddleBufferMultififoRegionizerEmulator::initSectorsAndRegions(const RegionizerDecodedInputs& in,
+                                                                          const std::vector<PFInputRegion>& out) {
+  assert(!init_);
+  init_ = true;
+  assert(out.size() == nregions_post_);
```
Since the amount of asserts here is relatively low, it would be probably better to throw `cms::Exception`s which can have some more information, than just an assert, which will only mention the line number of the failed assertion on failure.

gpetruc replied:
I can't use cms::Exception's since this code is also compiled outside of CMSSW in Vitis HLS testbenches.
I could use std exceptions, but eventually these asserts are not supposed to ever trigger when the code is running, they're there just to catch mistakes during development and to inform the reader of the assumptions that this piece of code is doing, so there's not much gain in making them more verbose.

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/src/regionizer/middle_buffer_multififo_regionizer_ref.cpp:
```diff
@@ -0,0 +1,632 @@
+  for (unsigned int ieta = 0; ieta < neta; ++ieta) {
+    for (unsigned int iphi = 0; iphi < nphi; ++iphi) {
+      unsigned int iin = ieta * nphi + iphi;
+      //if (ieta == 0 && iphi == 0 && pre_out_tk[iin].hwPt != 0) std::cout << "iclock " << iclock_ << ",  pre_out_tk[" << iin << "] = " << pre_out_tk[iin].pack().to_string(16) << std::endl;
```
Please remove the commented code

aloeliger commented:
Alright, I think I'm good here. @epalencia Anything from your side? If not, @gpetruc Could you go ahead and open a PR for this to the main CMSSW if there isn't one already?

epalencia commented:
Nothing from my side. @gpetruc , please go ahead and open the PR in master.

gpetruc commented:
Hi,

I noticed that when removing the debug printout I had introduced a formatting issue, so now I've re-run code-format and updated the last commit.

`L1Trigger/L1TTrackMatch/plugins/L1TrackJetClustering.h` also have formatting issues, I could add the fixes to it in this PR but I think it would be better if it's done in any of the other open GTT PRs (also to avoid potential conflicts).

aloeliger commented:
@gpetruc There's a number of accumulated formatting issues that I will likely fix by hand when your PR is merged. Don't worry about it for now. The IB is sort of a rough draft/best effort sort of thing after we've changed up how we use it.

gpetruc commented:
Hi,
Now the central CMSSW PR is signed, so maybe this can be merged.

epalencia commented:
@gpetruc , ORPs signature is still not there yet, would you mind to wait a bit for it or you need it already in the IB?

gpetruc commented:
I can wait a few dayxs, but eventually I'd like to have it in the IB since there's a large firmware PR that I can't merge until this one is merged.

epalencia commented:
Ok, if main PR is not merged in master tomorrow afternoon, I'll merge it here.

epalencia commented:
Tagged as `phase2-l1t-1330pre3_v12`.
